tephra-core/src/main/java/co/cask/tephra/visibility/VisibilityFence.java

```diff
@@ -0,0 +1,128 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.tephra.visibility;
+
+import co.cask.tephra.TransactionAware;
+
+/**
+ * VisibilityFence is used to ensure that after a given point in time, all readers see an updated change
+ * that got committed.
+ * <p>
+ *
+ * Typically a reader will never conflict with a writer, since a reader only sees committed changes when its
+ * transaction started. However to ensure that after a given point all readers are aware of a change,
+ * we have to introduce a conflict between a reader and a writer that act on the same data concurrently.
+ *<p>
+ *
+ * This is done by the reader indicating that it is interested in changes to a piece of data by using a read fence
+ * in its transaction. If there are no changes to the data when reader completes the transaction
+ * with the read fence, the commit succeeds.
+ * <p>
+ * After a writer is done updating the data that a reader is interested in, it tries to commit the write fence in a
+ * separate transaction.
+ * If the writer is able to successfully commit the transaction with write fence, it means that
+ * any in-progress readers that have not seen the change will not be allowed to commit anymore.
+ * In case an in-progress reader commits before the write fence is committed, then
+ * the writer will get a conflict on commit, and the writer has to retry the write fence with a new transaction
+ * until it successfully commits.
+ * <p>
+ * Any in-progress readers after the write fence is committed will get a conflict when they try to commit. This will
+ * force the reader to start a new transaction, and this ensures that the changes made by writer are visible
+ * to the reader.
+ * <p>
+ * Hence a successful commit of a write fence ensures that any read fence that successfully commits after
+ * this point onwards would have seen the change.
+ *
+ * <p>
+ * Sample reader code:
+ * <pre>
+ *   <code>
+ * TransactionAware readFence = VisibilityFence.reader(fenceId);
+ * TransactionContext readTxContext =
+ *        new TransactionContext(new InMemoryTxSystemClient(txManager), readFence, table1, table2, ...);
+ * readTxContext.start();
+ *
+ * // do operations using table1, table2, etc.
+ *
+ * // finally commit
+ * try {
+ *   readTxContext.finish();
+ * } catch (TransactionConflictException e) {
+ *   // handle conflict by aborting and starting over with a new transaction using the same fence object
+ * }
+ *   </code>
+ * </pre>
+ *<p>
+ *
+ * Sample writer code:
+ * <pre>
+ *   <code>
+ * // start transaction
+ * // write change
+ * // commit transaction
+ *
+ * // Now use a write fence (with the same fenceId) to ensure that all in-progress readers are aware of this change
+ * TransactionAware writeFence = VisibilityFence.writer(fenceId);
```

Comment:
I think the API should be more intuitive such that the user doesn't need to deal with the retry. Something like:

```java
try {
  VisibilityFence.await(txContext, fence, 5L, TimeUnit.SECONDS);
} catch (TimeoutException e) {
  // Handle timeout case.
}
```

The retry will be handled inside the await with some sleep. We can also have an advanced version that looks like:

```java
void await(TransactionContext context, RetryStrategy retryStrategy) throws SomeMeaningfulException;
```